jplag/src/main/java/de/jplag/JPlagComparison.java

```diff
@@ -74,37 +76,19 @@ public boolean equals(Object other) {
         }
 
         TokenList tokenList = (j == 0 ? firstSubmission : secondSubmission).getTokenList();
-        int i, h, starti, starth, count = 1;
-
-        o1: for (i = 1; i < matches.size(); i++) {
-            starti = (j == 0 ? matches.get(i).getStartOfFirst() : matches.get(i).getStartOfSecond());
-            for (h = 0; h < i; h++) {
-                starth = (j == 0 ? matches.get(h).getStartOfFirst() : matches.get(h).getStartOfSecond());
-                if (tokenList.getToken(starti).file.equals(tokenList.getToken(starth).file)) {
-                    continue o1;
-                }
-            }
-            count++;
-        }
 
-        String[] res = new String[count];
-        res[0] = tokenList.getToken((j == 0 ? matches.get(0).getStartOfFirst() : matches.get(0).getStartOfSecond())).file;
-        count = 1;
-
-        o2: for (i = 1; i < matches.size(); i++) {
-            starti = (j == 0 ? matches.get(i).getStartOfFirst() : matches.get(i).getStartOfSecond());
-            for (h = 0; h < i; h++) {
-                starth = (j == 0 ? matches.get(h).getStartOfFirst() : matches.get(h).getStartOfSecond());
-                if (tokenList.getToken(starti).file.equals(tokenList.getToken(starth).file)) {
-                    continue o2;
-                }
-            }
-            res[count++] = tokenList.getToken(starti).file;
+        /*
+         * Collect the file names of the first token of each match.
+         */
+        Set<String> collectedFiles = new LinkedHashSet<>();
+        for (Match match: matches) {
+            collectedFiles.add(tokenList.getToken(match.getStart(j == 0)).file);
         }
```

Comment:
This is quite an improvement code-style-wise. Just to have some sort of sanity test I will write a quick unit test to ensure identical functionality.

Reply:
Works as expected!

Reply:
I ran a test by checking lengths of both results before deleting the old `count` computation, should perhaps have added that.